exercises/tree_iterators/descr.md

```diff
@@ -0,0 +1,180 @@
+# Tree Iterators
+
+In this exercise, we study how to **produce (on demand) a sequence of the
+elements of a data structure**. We use a simple binary-tree data structure as
+a representative example, but our technique is applicable to any data
+structure.
+
+We use the type `'a Seq.t` of **on-demand sequences** to represent a producer
+of a sequence of elements. The main point that we wish to make is that **this
+type is very similar to the type of lists**, `'a list`, so implementing a
+producer of a sequence of elements requires essentially the same programming
+skill as building a list of elements.
+
+## Binary trees
+
+A type `'a tree` of binary trees is given, where every binary node is
+decorated with an *element*, that is, a piece of data of type `'a`.
+This type is defined as follows:
+
+```
+type 'a tree =
+  | Leaf
+  | Node of 'a tree * 'a * 'a tree
+```
+
+One could impose invariants on this data structure, including an ordering
+invariant and a balancing invariant, to obtain *balanced binary search trees*,
+a simple, widely-known implementation of dictionaries. In this exercise,
+however, these invariants play no role.
+
+The **fringe** of a tree is the list of its elements, in the order in which
+they appear when the tree is displayed as an OCaml expression. (This is
```

Comment:
This definition seems a bit fragile and underspecified. What does "displayed as an OCaml expression" mean? "displayed" by who? The toplevel? A custom printer? Sticking with the usual "infix enumeration" seems better to me.

Reply:
The toplevel puts ellipsis when it prints large values:
```
let rec big n = if n = 0 then Leaf else Node (big (n - 1), n, big (n - 1));;
```
Typically, `big 10` is displayed with such ellipsis.

Reply:
Ok, I have replaced "displayed" with "written down". I like to say things this way because I believe "infix enumeration" is not easily understood by a beginner. Furthermore, I have given an example, so I think my definition of **fringe** is acceptable. I can't give a formal definition anyway, since this is the topic of Question 1.